src/Microsoft.AspNet.Razor/SourceLocation.cs

```diff
@@ -45,30 +75,49 @@ public override string ToString()
                 CharacterIndex);
         }
 
+        /// <inheritdoc />
         public override bool Equals(object obj)
         {
             return (obj is SourceLocation) && Equals((SourceLocation)obj);
         }
 
+        /// <inheritdoc />
         public override int GetHashCode()
         {
             // LineIndex and CharacterIndex can be calculated from AbsoluteIndex and the document content.
-            return AbsoluteIndex;
+            return HashCodeCombiner.Start().Add(FilePath).Add(AbsoluteIndex).CombinedHash;
         }
 
+        /// <inheritdoc />
         public bool Equals(SourceLocation other)
         {
-            return 
+            return string.Equals(FilePath, other.FilePath, StringComparison.OrdinalIgnoreCase) &&
                 AbsoluteIndex == other.AbsoluteIndex &&
                 LineIndex == other.LineIndex &&
                 CharacterIndex == other.CharacterIndex;
         }
 
+        /// <inheritdoc />
+        /// <exception cref="ArgumentException">if the <see cref="FilePath"/> of the left and right operands
+        /// are different.</exception>
         public int CompareTo(SourceLocation other)
         {
+            if (!string.Equals(FilePath, other.FilePath, StringComparison.OrdinalIgnoreCase))
```

Comment:
I guess if we don't necessarily care about the semantics of the Comparison, just the fact that it doesn't throw - I'll use `StringComparison.Ordinal` here.